cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -381,13 +381,21 @@ Result<S3Options> S3Options::FromUri(const std::string& uri_string,
 }
 
 bool S3Options::Equals(const S3Options& other) const {
+  const int64_t default_metadata_size = default_metadata ? default_metadata->size() : 0;
+  const bool default_metadata_equals =
+      default_metadata_size
```

Comment:
nit: is it possible to have a non-nullptr `default_metadata` that is of size 0? Should this check instead be `default_metadata`?